### ISSUE TITLE
Truncation works slowly. Tokenizer firstly encodes long sequence and then truncate it.

### ISSUE THREAD

**Issue by galtimur** [huggingface/tokenizers#1573]
I work with very long sequences, which I want to truncate to a shorter one in the pre-processing stage.
I found that the Tokenizer first encodes the whole sequence whole, after that truncates it. Is there a way to truncate it on the fly, to save time on tokenization of unused part of the sequence?

I use transformers 4.42.4.
tokenizer class is LlamaTokenizerFast

The code and the output:

```
from transformers import AutoTokenizer
from time import time

tokenizer = AutoTokenizer.from_pretrained("deepseek-ai/deepseek-coder-1.3b-base")
# tokenizer class is LlamaTokenizerFast
def encode_long(meta_len, batch_size=200):
    text = 500*"jibber-jabber"
    text_batch = batch_size*[meta_len*text]

    start_time = time()
    batch = tokenizer.batch_encode_plus(text_batch, truncation=True, max_length=1000)
    print(f"Seq encoded len = {len(batch['input_ids'][0])}; Time used = {time() - start_time:.2f}, Symbols = {meta_len*len(text)}")

encode_long(meta_len=1)
encode_long(meta_len=10)
encode_long(meta_len=100)
```

Results of the code on my machine:

Seq encoded len = 1000; Time used = 0.07, Symbols = 6500
Seq encoded len = 1000; Time used = 0.59, Symbols = 65000
Seq encoded len = 1000; Time used = 8.98, Symbols = 650000

[Colab ](https://colab.research.google.com/drive/1HkEBto8wgL4MbBp4dqZJ0hH_6KNUS1Ka?usp=sharing) for reproduction.

**Comment by ArthurZucker** [comment]
Hey! There is currently no way to do that because what you want sounds more like truncate the text directly no? 
We are gonna improve the throughput in general, but you need to encode first, otherwise what you need is just cutting your texts to the correct length no?

**Comment by galtimur** [comment]
> Hey! There is currently no way to do that because what you want sounds more like truncate the text directly no?

Yes, that's true. I overcome this issue by first truncating the text directly. To do so I have to assess the average length of the token with some extra margin.
I thought that it would be natural to truncate sequences right in the tokenization loop, breaking it when the desired number of tokens is reached.

**Comment by komitydev** [comment]
i've encountered the same need for "early stopping" during the tokenization. is that a hard feature to implement? i find it very useful, strange that is hasn't been done earlier. because average token-char length is not precise and can work well enough, but still early stopping feature will be a big optimization step for the library

**Comment by ArthurZucker** [comment]
No it's not hard at all! 
We can do many different things, but most probably a `pre_tokenizer`. I'll open a PR but I am not entirely sure about the API you'd want. Replacing the current truncation might be feasible.
